storage-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java

```diff
@@ -57,11 +155,167 @@ public Boolean isPartialWriteVisible(Path path, Configuration hadoopConf) throws
+        copyFile(fs, src, dst);
+    }
+
+    protected void writePutCompleteDbEntry(ExternalCommitEntry entry) throws IOException {
+        putExternalEntry(entry.asComplete(), true);
+    }
+
+    protected void fixDeltaLogCopyTempFile(FileSystem fs, Path src, Path dst) throws IOException {
+        copyFile(fs, src, dst);
+    }
+
+    protected void fixDeltaLogPutCompleteDbEntry(ExternalCommitEntry entry) throws IOException {
+        putExternalEntry(entry.asComplete(), true);
+    }
+
     ////////////////////
     // Helper Methods //
     ////////////////////
+
+    /**
+     * Method for assuring consistency on filesystem according to the external cache.
+     * Method tries to rewrite TransactionLog entry from temporary path if it does not exists.
+     * Method returns completed [[ExternalCommitEntry]]
+     */
+
+    private void fixDeltaLog(FileSystem fs, ExternalCommitEntry entry) throws IOException {
+        if(entry.complete) {
+            return;
+        }
+        int retry = 0;
+        boolean copied = false;
+        while(true) {
+            LOG.debug(String.format("trying to fix: %s", entry.fileName));
+            try {
+                if (!copied && !fs.exists(entry.absoluteJsonPath())) {
+                    fixDeltaLogCopyTempFile(fs, entry.absoluteTempPath(), entry.absoluteJsonPath());
+                    copied = true;
+                }
+                fixDeltaLogPutCompleteDbEntry(entry);
+                LOG.info(String.format("fixed %s", entry.fileName));
+                return;
+            } catch(Throwable e) {
+                LOG.info(String.format("%s: %s", e.getClass().getSimpleName(), e));
+                if(retry >= 3) {
+                    throw e;
+                }
+            }
+            retry += 1;
+        }
+    }
+
+   /**
+    * Copies file within filesystem
+    * @param fs reference to [[FileSystem]]
+    * @param src path to source file
+    * @param dst path to destination file
+    */
+    private void copyFile(FileSystem fs, Path src, Path dst) throws IOException {
+        LOG.debug(String.format("copy file: %s -> %s", src, dst));
+        FSDataInputStream input_stream = fs.open(src);
+        FSDataOutputStream output_stream = fs.create(dst, true);
+        try {
+            IOUtils.copy(input_stream, output_stream);
+            output_stream.close();
+        } finally {
+            input_stream.close();
+        }
+    }
+
+    /**
+     * Returns path stripped user info.
+     */
+    private Path stripUserInfo(Path path) throws java.net.URISyntaxException {
+        URI uri = path.toUri();
+        URI newUri = new URI(
+            uri.getScheme(),
+            null,
+            uri.getHost(),
+            uri.getPort(),
+            uri.getPath(),
+            uri.getQuery(),
+            uri.getFragment()
+        );
+        return new Path(newUri);
+    }
+
+    // TODO - use java version of FileNames utils?
```

Comment:
this has been added in this PR here: https://github.com/delta-io/delta/pull/995

So we can refactor once #995 is merged.